v2x_ros_driver/src/v2x_radio_client.cpp

```diff
@@ -168,29 +169,138 @@ void V2XRadioClient::process(const std::shared_ptr<const std::vector<uint8_t>>&
         if ((i + 1 + len + len_bytes) < entry.size()) {
-            // bool found_valid_msg = true;
             if (!IsValidMsgID(std::to_string(msg_id))) { continue; }
+
             size_t start_index = i;
             size_t end_index = i + 2 + len + len_bytes; // includes 2 msgID bytes before message body
-            //this constructor has range [first, last) hence the + 1
+            // this constructor has range [first, last) hence the + 1
             std::vector<uint8_t> msg_vec(entry.begin() + start_index, entry.begin() + end_index);
-            onMessageReceived(msg_vec, msg_id);
-            break;
+
+            // Check if specified MessageFrame size matches actual data size
+            printVector(msg_vec);
+            if (!isValidMsgSize(msg_vec, start_index, end_index, entry))
+            {
+                continue;
+            }
+            else
+            {
+                if (!isValidPSID(std::to_string(msg_id)))
+                {
+                    printVector(msg_vec);
+                    onMessageReceived(msg_vec, msg_id);
+                    break;
+                }
+                else
+                {
+                    continue;
+                }
+            }
         }
     }
 }
 
+void V2XRadioClient::printVector(const std::vector<uint8_t>& vec)
+{
+    std::ostringstream oss;
+    oss << "[";
+    for (size_t i = 0; i < vec.size(); ++i) {
+        oss << static_cast<int>(vec[i]);
+        if (i < vec.size() - 1) {
+            oss << ", ";
+        }
+    }
+    oss << "]";
+    RCLCPP_INFO_STREAM(logger_, "Sending message vector: " << oss.str());
+}
+
+std::vector<uint8_t> V2XRadioClient::hexStringToByteArray(const std::string& hexString) const
+{
+    std::vector<uint8_t> byteArray;
+
+    // Loop through the hex string, two characters at a time
+    for (size_t i = 0; i < hexString.length(); i += 2) 
+    {
+        // Extract two characters representing a byte
+        auto byteString = hexString.substr(i, 2);
+
+        // Convert the byte string to a uint8_t value
+        auto byteValue = static_cast<uint8_t>(stoi(byteString, nullptr, 16));
+
+        // Add the byte to the byte array
+        byteArray.push_back(byteValue);
+    }
+
+    return byteArray;
+}
+
+bool V2XRadioClient::isValidMsgSize(const std::vector<uint8_t> msg_vec, size_t start_index, size_t end_index, const std::vector<uint8_t> entry)
+{
+    if (msg_vec.size() > 255) 
+    {
+        auto tmp_start_index = start_index + long_frame_;
+        std::vector<uint8_t> long_vec(entry.begin() + tmp_start_index, entry.begin() + end_index);
+        auto msg_size = (static_cast<uint16_t>(msg_vec[2]) << 8) | msg_vec[3];
+        if (msg_size == long_vec.size())
+            {
+                return true;
+            }
+        else 
+        {
+            RCLCPP_WARN_STREAM(logger_, "Size in possible MessageFrame does not match actual data size. Checking rest of data.");
```

Comment:
I think we need to move all or part of this log to the calling function. This log states "Checking rest of data." but `isValidMsgSize()` does not do anymore checks, they're handled in `process()`

Reply:
Updated